### PR TITLE
CLI: Detach automigrate command from storybook init

### LINKED ISSUE #22346: storybook init: Detach automigrate from init
Error location: [automigrate](https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/automigrate/index.ts#L92)

Suggested work:

This is a bug where the `fixes` object from running automigration is not an object. It seems to be related to [these](https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/automigrate/index.ts#L160) [early](https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/automigrate/index.ts#L175) [bails](https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/automigrate/index.ts#L186) when running automigrations that could happen due to the CLI:

- not detecting main.js (user has to run the command again with —config-dir, but it does not make sense when running sb init)
- having errors trying to parse the main.js file (maybe something else failed and there is no main.js)

This happens because we always run automigrate as part of init, and it can cause potential issues for users without necessity for that.

We currently run automigrations as part of init because for select packages, typescript (or some babel features) are not supported out of the box: 

```groovy
@storybook/react-webpack5
@storybook/vue-webpack5
@storybook/vue3-webpack5
@storybook/html-webpack5
@storybook/web-components-webpack5
```

As part of this task, we should detach automigrate from init, and provide a better solution for the projects listed above to work out of the box, such as running the code from `missing-babelrc` automigration as part of the init step, only for that select group of frameworks.

The other reason we run automigrations is to provide users the option to install `eslint-storybook-plugin` automatically, which is really helpful. We should also move this suggestion to the sb init command.

### PR DESCRIPTION
**Pull request by yannbf**
Closes #22346 

## What I did

- Automigrate does not run as part of `storybook init` anymore
- init command automatically detects and adds a babel config if needed (similar to the missing-babelrc automigration, but automated with no prompts)
- old ESLint "no-anonymous-default-exports" logic is removed as it was legacy thing for CRA and it's handled in the Storybook ESLint plugin
- eslint plugin logic was improved to also support package.json settings
- eslint plugin logic was refactored and reused in two places: 
  1. automigrations (as it existed already) for users who already have Storybook and are upgrading
  2. init (new) where it will suggest users to install the ESLint plugin as part of initializing Storybook

image

Text formatting isn't amazing but we plan on revamping the CLI to use something else and more beautiful in the future, I guess we can cope with it for now!

## How to test

1. Build the CLI
2. Go to different projects that do not have Storybook (use sandboxes repo) and use the `storybook init` command in those:
  a. [react-webpack](https://github.com/storybookjs/sandboxes/tree/next/react-webpack/17-ts/before-storybook) - should add babel config file
  b. [nextjs](https://github.com/storybookjs/sandboxes/tree/next/nextjs/default-js/before-storybook) - should suggest ESLint plugin and change the eslint config file
  c. [cra](https://github.com/storybookjs/sandboxes/tree/next/cra/default-js/before-storybook) - should suggest ESLint plugin and change config in package.json

## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: code/lib/cli/src/generators/baseGenerator.ts [hunk 1]
```diff
@@ -260,20 +263,60 @@ export async function baseGenerator(
...
+      await writeBabelConfigFile({
+        typescript: isTypescript,
+        jsx: isReact,
+      });
+    }
   }
 
-  const depsToInstall = [...versionedPackages, ...babelDependencies];
+  try {
+    if (process.env.CI !== 'true') {
```
Comment (shilman) [inline code/lib/cli/src/generators/baseGenerator.ts:301]: In all other places in the code we just check for `if(process.env.CI)`

File: code/lib/cli/src/babel-config.ts [hunk 2]
```diff
@@ -52,48 +101,29 @@ export const generateStorybookBabelConfig = async ({ target }: { target: string
...
 
   const { runInstall } = await prompts({
     type: 'confirm',
     initial: true,
     name: 'runInstall',
-    message: `Shall we install the required dependencies now? (${added.join(', ')})`,
+    message: `Shall we install the required dependencies now? (${dependencies.join(', ')})`,
   });
 
   if (runInstall) {
```
Comment (shilman) [inline code/lib/cli/src/babel-config.ts:116]: `runInstall` not only installs the dependencies but adds them to `package.json`. That means that if you don't install them, you'll be in an inconsistent state where the config file has been generated but even running `npm install` won't get you the deps you need. This is already warned about below, but it's a weird behavior and I think even if the user opts out of installing the deps they should still get added to `package.json.

Comment (yannbf) [reply code/lib/cli/src/babel-config.ts:116]: I didn't change the behavior of this code, but I could try and make this change if you like!

Some context:
- This code is NOT executed as part of `storybook init` – the babel generation is handled differently and the dependencies are installed correctly
- This code is executed as part of `automigrate`
- This code is executed when users run `storybook babelrc`

Normally we only add dependencies to the package.json without installing when we know exactly which version to put there (e.g. from versions.ts), but in this case, if we were to skip installing, given that we have no idea which version to put, we'd need to either fetch the latest version using their package manager or add `"latest"` to package.json.

Optionally, we could just run install without prompting the user

File: code/lib/cli/src/generators/baseGenerator.ts (hunk 1 above)
Comment (yannbf) [reply code/lib/cli/src/generators/baseGenerator.ts:301]: If we do that, the if condition will be matched even if users use `CI=false`, as the value of CI will become `"false"`, and therefore will be truthy 🤔

File: code/lib/cli/src/babel-config.ts (hunk 2 above)
Comment (shilman) [reply code/lib/cli/src/babel-config.ts:116]: Yup, I saw that. Do as you see fit!

Comment (yannbf) [reply code/lib/cli/src/babel-config.ts:116]: Removed the prompt! Thanks for catching this! I think it's better now

File: code/lib/cli/src/automigrate/helpers/eslintPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  const extension = SUPPORTED_ESLINT_EXTENSIONS.find((ext: string) =>
+    fse.existsSync(`${filePrefix}.${ext}`)
+  );
+  return extension ? `${filePrefix}.${extension}` : null;
+};
+
+export async function extractEslintInfo(packageManager: JsPackageManager) {
+  const allDependencies = await packageManager.getAllDependencies();
+  const packageJson = await packageManager.retrievePackageJson();
+  let eslintConfigFile;
```
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/helpers/eslintPlugin.ts:33]: Could you add some typings here?

File: code/lib/cli/src/automigrate/helpers/eslintPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    });
+  }
+}
+
+export const suggestESLintPlugin = async (): Promise<boolean> => {
+  const { shouldInstall } = await prompts({
+    type: 'confirm',
+    name: 'shouldInstall',
+    message: dedent`
+        We detected that you're using ESLint. Storybook provides a plugin that gives the best experience with Storybook and helps follow best practices: ${chalk.yellow(
```
Comment (valentinpalkovic) [inline code/lib/cli/src/automigrate/helpers/eslintPlugin.ts:85]:
```suggestion
        We have detected that you're using ESLint. Storybook provides a plugin that gives the best experience with Storybook and helps follow best practices: ${chalk.yellow(
```

**Review by valentinpalkovic — APPROVED** [review]

File: code/lib/cli/src/automigrate/helpers/eslintPlugin.ts (hunk 3 above)
Comment (yannbf) [reply code/lib/cli/src/automigrate/helpers/eslintPlugin.ts]: Sure thing, please check!

**Review by valentinpalkovic — APPROVED** [review]

**Pull request merged**